### DISCUSSION TITLE
Allow invoking readonly members on structs in object initializers

### DISCUSSION THREAD

**Discussion by TahirAhmadov** [dotnet/csharplang#8427] [Language Ideas] [answered]
Imagine you have a lazy-instantiated collection, and a "proxy collection" to wrap around it:
```cs
using System;
using System.Collections;
using System.Collections.Generic;

var foo = new Foo
{
  Strings = { "abc" },
  Numbers = { 123 } // currently, this is CS1918
};
foo.Numbers.Add(456);
Console.WriteLine(string.Join(", ", foo.Strings));
// output: abc, 123, 456

class Foo
{
  public IList<string> Strings => this._list ??= [];
  public Wrapper Numbers => new Wrapper(ref this._list);

  List<string>? _list;
}
public readonly ref struct Wrapper : IEnumerable // currently, this is CS8343, but this will soon be possible
{
  public Wrapper(ref List<string>? list)
  {
    this._list = ref list;
  }
  public readonly void Add(int i)
  {
    (this._list ??= []).Add(i.ToString());
  }

  readonly ref List<string>? _list;

  //public IEnumerator<int> GetEnumerator() => throw new NotSupportedException();
  IEnumerator IEnumerable.GetEnumerator() => throw new NotSupportedException();
}
```

The gist of it is, the CS1918 shouldn't be generated, because the `Wrapper.Add` is `readonly`, therefore it's safe to invoke it as part of collection initializer. CS1918 should continue to be generated in all other existing cases.

**Comment by jaredpar** [comment]
This seems like a case for `init` on methods not changing the meaning of `readonly`. The use case you want here is essentially exactly what `init` provides today.

**Reply by TahirAhmadov** [reply]
I'm not sure what `init` has to do with this, can you please clarify?

**Reply by jaredpar** [reply]
The core behavior you want to achieve is to allow mutation of the object during the initializer but for it to be read only afterwards. This is exactly what [init property accessors](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/init.md) set out to solve. They are accessible in the ctor, other `init` members and object initializers. 

The idea of `init` doesn't have to apply to only property setters. It's could also be extended to apply to method as well. This [was considered](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-9.0/init.md#init-members) when we did `init` accessors but decided to wait until we had more reason for need. 

This seems like the most likely route to get this behavior.

**Reply by jaredpar** [reply]
Hmm ... missed a bit that you're using a `ref` field here. That changes it a bit.

**Reply by TahirAhmadov** [reply]
@jaredpar  The objective here is to twofold: 1) allow making a struct typed property as a proxy for something else, like in this particular scenario; 2) bring the struct manipulation rules in object initializers more inline with rules outside of object initializers. I don't think it has anything to do with `init` keyword.

**Comment by timcassell** [comment]
This code doesn't make sense to me. Numbers property is a getter for a ref struct. How would you even be able to write that to a class? It's illegal even without CS1918.

It seems that you would want to have a collection expression work by adding elements to an existing collection, rather than creating a new one. I don't think that even makes sense with the assignment operator, and maybe needs something like a backward pipe operator `<|`?

**Reply by TahirAhmadov** [reply]
[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8CwAUAAIBMAjEcQAwAEx5AdAMID2ANuwKYDGCUrGAGcA3FToMALGMJEAbhGS0AZq1a0AvLRhcA7rQBiaogG8itWgGUEyWAHMhm2idoAiCMB6vaAXzjnaADlUMGAuZEctF3JSAGZfWgB6RNoeVGRkLhgEdgBPOFoEAAsoR1LaZktyAE5yAA4iHxkGaoAKBmpGAClWWFbXAoGVNUZrWxgHAEpJmWTaVlQEAAdFkFoPHgKY2KIqUkNjQjNCC2J4gEkAGVKEAB4OgD4rG3tIp+LSxgB9dhvaAH5/loANoAXRkp3iAHVkBAlktwkEQmEIponjp9DC4QjkK1MspCiUhN9fkIEDNdidaNcyfdyNQHv9aD8bjIfFR4pkIBhBHlaPjaGS0HxaFj4Yi1gxYrdYAgnnM0hksjl8oTyuVKnVYpJYgVgIs1Y5dFBOIK1DBaGFaEtWEIhFBgNxTAEzqLYeLcQKaXdHrRSeSAscLBYPsSWWSnP6Ib4XZyuNzebl6JJaABBDAYVqy2hQSaBgIWVqhkl/QEg0GTRjpzNQRgAFVYY3srWm0fZsipXJ5MD5XpudIZTPDCBksdo5wAojAQuEIAhWMgZdkngBxLgIKcz2Hz3GTNGE5CsfQYoKsBCWVDwhcILgYCcADx4XCW/EELejk+nYFnO/o5FijBrhuX4/guLb7sUh7Hnop7npeNrIDed6Ps+r4wO+jREEAA==)

The only error being reported is the CS1918.

**Reply by TahirAhmadov** [reply]
PS. I'm not sure what collection expressions have to do with this, and also what `<|` is supposed to do, any details would be appreciated.

**Reply by timcassell** [reply]
When it comes to compiler errors, it doesn't always report all possible errors. Fix one error and another can appear. Rinse and repeat until all errors are fixed.

> I'm not sure what collection expressions have to do with this, and also what <| is supposed to do, any details would be appreciated.

`Numbers = { 123 }` syntax is the old collection initializer syntax, which can now be written using collection expression syntax `Numbers = [ 123 ]`. To me, it doesn't make sense in this context to assign it, but rather to pipe it `Numbers <| [ 123 ]`. Even if you did actually want to assign it, the collection initializer uses the empty params constructor, which gets you a null ref, so it would explode at runtime anyway.

Piping in this case would call `Add` for each element on the existing instance, instead of creating a new instance.

**Reply by TahirAhmadov** [reply]
If I remove the CS1918 offending line, it compiles fine, and works in non-preview on my machine, too. Returning `ref struct` values from a property is legal.

[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8CwAUAAIBMAjEcQAwAEx5AdAMID2ANuwKYDGCUrGAGcA3FToMALGMJEAbhGS0AZq1a0AvLRhcA7rQBiaogG8itWgGUEyWAHMhm2idoAiCMB6vaAXzjnaAHpAgDlUMGAuZEctF3JSAGZfIMDaHlRkZC4YBHYATzhaBAALKEcy2mZLcgBOcgAOIh8ZBhqACgZqRgApVlg210LBlTVGa1sYBwBKKZlg2lZUBAAHJZBaDx5C+ISiKlJDY0IzQgtiJIBJABkyhAAeToA+Kxt7GOeSssYAfXZb2gA/ACtABtAC6MjOSQA6sgIMtllFaGEIlF3to9LRYfDEcg2lllEVSkIfn8hAhZntTrQbuSHuRqI8AbRfrcZD4qEkshAMIJ8rQCbRyWg+Fi4QikesGAk7rAEM95ulMtlcgUiRUKlV6glJAlCsAlurHLooJwhWoYLRIrRlqwhEIoMBuKYAucxTioviuITafcnrQyRSAicLBZPiTWeSnIHIb5XVyuDy+Xl6JJaABBDAYNpy2hQKbBgIWNrh0n/IGgsFTRiZ7NQRgAFVY43sbRmsY5smp3N5MH5gt99MZzMjCBk8doFwAojBwlEIAhWMhZTlngBxLgIGdzuGLvFTTQfYrIVj6HT6EKsBCWVAIpcILgYKcADx4XGW/EEbdj09nYHne70OQCSMBuW5/gBS5toeRInmemKXtet62sgD5Pq+76fjA35NEQQA)

Regarding the collection expression, the `Numbers = [ 123 ]` has different semantics than `Numbers = { 123 }`. 
I had proposed `Numbers += [ 123 ]` syntax (because I also don't like curly braces mixed with collections), which is kind of what your proposed `<|` does, but frankly I think my proposal for `Numbers += [ 123 ]` has a higher chance of being added than `<|`, because 1) it combines existing language features, and 2) it allows spreads (which can call `AddRange` on compatible collections). 
However, all of this is irrelevant because the objective here is expose a zero-allocation proxy-type property to allow add values to a lazily-instantiated collection, and even if/when we get either `Numbers += [ 123 ]` or `Numbers <| 123`, we still need the improvement in my OP. (Unless they special case it somehow.)

> Even if you did actually want to assign it, the collection initializer uses the empty params constructor, which gets you a null ref, so it would explode at runtime anyway.

PS. This is also incorrect, because the `Number` property in question is get only - nothing is assigned, therefore no empty structs are passed around whatsoever. (BTW I hate the "empty params constructor" expression for describing `initobj` opcode - structs without an explicit parameterless ctor don't have one, period - the `new S()` is just a different way to write `default(S)` in those cases.)

**Reply by TahirAhmadov** [reply]
Corrected [SharpLap](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8CwAUAAIBMAjEcQAwAEx5AdAMID2ANuwKYDGCUrGAGcA3FToMALGMJEAbhGS0AZq1a0AvLRhcA7rQBiaogG8itWgGUEyWAHMhm2idoAiCMB6vaAXzjnaAHpAgDlUMGAuZEctF3JSAGZfIMDaHlRkZC4YBHYATzhaBAALKEcy2mZLcgBOcgAOIh8ZVVZGMIiooUYAQQwMAAp4hIBKGQYagYZqRgApVlgB10LllTVGa1sYBxGxomDaVlQEAAdjkFoPHkLhoipSQ2NCM0ILYiSASQAZMoQAHmmAD4rDZ7DFgSUyowAPrsX60AD8CK0AG0ALoyN5JADqyAgJxOUVoHUi0U0wJ0+lx+MJyAGWWURVK3Vhvz2sletB+Qn+QIRtFZPJkPioSSyEAwgnytAZtB5aD4tGpBKJFwYSQO6Uy2VyBSZFQqVXqCUkCUKwGO+scuignDlahgtEitBOrCEQigwG4pgC7yVeJVdNl3N55GogP5cJ5IwCLwsFkhLKjCCcsuTmN8vrFXAlUry9EktD6g1gKagMc5cfjA0TMOTiORtHRI16/QGUEYABVWJt7ANdhmRRyLOLJTBpcHfgCwxGBem7pzgn6PgBRGDhKIQBCsZB/UvAgDiXAQa43eO3dJG5KZyFY+kpxNYCEsqAJO4QXAwK4AHjwuCd+EEfsM1XdcwE3C96HIBJGCPE8wIgnd+2vEpb3vPRH2fV9XWQD8v1/f9AJgYCmiIIA===)

**Reply by timcassell** [reply]
> Returning `ref struct` values from a property is legal.

Yes, I know. But your code is trying to write it, not read it. 

> Regarding the collection expression, the `Numbers = [ 123 ]` has different semantics than `Numbers = { 123 }`. I had proposed `Numbers += [ 123 ]` syntax (because I also don't like curly braces mixed with collections), which is kind of what your proposed `<|` does, but frankly I think my proposal for `Numbers += [ 123 ]` has a higher chance of being added than `<|`, because 1) it combines existing language features, and 2) it allows spreads (which can call `AddRange` on compatible collections). 

Yes, `+=` operator could also work. That was my point that a simple `=` cannot work. 

> However, all of this is irrelevant because the objective here is expose a zero-allocation proxy-type property to allow add values to a collection, and even if/when we get either `Numbers += [ 123 ]` or `Numbers <| 123`, we still need the improvement in my OP. (Unless they special case it somehow.)

Well, one feature is dependent on the other if they don't special-case it. But removing CS1918 error could technically be done separately if there is need for it.

[Edit] Actually there is no need to change that error as you can use `init` as @jaredpar mentioned. 

> This is also incorrect, because the `Number` property in question is get only - nothing is assigned, therefore no empty structs are passed around whatsoever.

But as I already mentioned, you're using the assignment operator in your example.

**Reply by TahirAhmadov** [reply]
> Yes, I know. But your code is trying to write it, not read it.
> Yes, += operator could also work. That was my point that a simple = cannot work.
> But as I already mentioned, you're using the assignment operator in your example.

No no. The `Numbers = { 123 }` does **not** assign; it does not write, it reads only. It's not a simple `=`.

> Well, one feature is dependent on the other if they don't special-case it. But removing CS1918 error could technically be done separately if there is need for it.

I can imagine a use case for a struct-type proxy property which does other things, not just collection addition.
```cs
class Bar
{
  public string StrProp { get; set; }
  public Proxy Proxy => new Proxy(this);
}
struct Proxy
{
  internal Proxy(Bar bar) => this._bar = bar;
  public int IntProp { set => this._bar.StrProp = value.ToString(); }
}
var bar = new Bar
{
  StrProp = "abc",
  Proxy = { IntProp = 123 }, // this overwrites StrProp - intentionally contrived example
};
```

**Reply by TahirAhmadov** [reply]
> [Edit] Actually there is no need to change that error as you can use init as @jaredpar mentioned.

He hasn't explained how, so I still have no idea what you two are talking about. Can you provide a code example?

**Reply by timcassell** [reply]
> No no. The `Numbers = { 123 }` does **not** assign; it does not write, it reads only. It's not a simple `=`.

Well, apparently I'm confused and this is a language feature I don't fully understand. Can you link me to docs about this feature? I'm not even sure what it's called. I thought it was equivalent to `Numbers = new Wrapper() { 123 }`.

**Reply by TahirAhmadov** [reply]
At your service:

https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/object-and-collection-initializers#object-initializers-with-collection-read-only-property-initialization

**Reply by TahirAhmadov** [reply]
The `Numbers = { 123 }` today does the same thing as `Numbers += [ 123 ]` or `Numbers <| [ 123 ]` would do in the future. The important thing here is that regardless which one of the 3 aforementioned syntaxes we use, it's still invoking `Wrapper.Add`, which is currently prohibited because `Wrapper` is a struct, which is the restriction I want to lift in cases where the member - `Add` in this case - is `readonly`.

**Reply by timcassell** [reply]
Thank you. So that syntax essentially takes an existing object and re-initializes it without instantiating a new object.

So the problem here it seems is that restriction exists for structs because modifying a struct returned from a property is working on a copy most of the time. Modifying copied structs usually doesn't propagate those changes, which makes sense why it's restricted in object initializers, as it can lead to subtle bugs.

In this particular case, modifying the struct _does_ propagate those changes. Perhaps the restriction could simply be lifted for `ref structs`. I can certainly imagine use-cases for regular structs, but it's harder to prove that the modifications will be observable. At least, the error could be demoted to a warning and let the user figure it out.

**Reply by TahirAhmadov** [reply]
Right, and that's why I think `readonly` is the correct criterion - by definition the struct doesn't get modified by such members.

**Reply by timcassell** [reply]
I think modifying public ref fields/properties should also be legal (not ref re-assign).

**Reply by TahirAhmadov** [reply]
> Modifying copied structs usually doesn't propagate those changes, which makes sense why it's restricted in object initializers, as it can lead to subtle bugs.

Funniest part is that doing this outside of object initializers is perfectly fine today, and it does introduce subtle bugs, but Cyrus was adamant that he doesn't want to make that illegal, even a warning.
So the same behavior is completely legal to a fault in one case, and completely illegal in another case. To me it looks like different people answered the same question radically differently - one of those decisions was made 25 years ago, the other much more recently (I can't tell for sure when, though.)
I think both of these approaches are incorrect; the `readonly` members should drive whether this kind of invocation should be legal or not, in both object initializers and elsewhere.
PS. Please see our glorious debate on this subject in https://github.com/dotnet/csharplang/discussions/8355.

**Reply by timcassell** [reply]
Ah yes, I even contributed a bit to that discussion. 😅 I agree with you. But in their defense, it's much more palatable to relax restrictions than to add them.

**Reply by TahirAhmadov** [reply]
Let's see if you are right about relaxing restrictions. I brought this discussion to Fred's attention, so far not a word from any of them. Perhaps they're thinking :)

**Reply by 333fred** [reply]
I mean, you got Jared thinking about this. I personally do not think this is a particularly compelling use case.

**Reply by TahirAhmadov** [reply]
Yes but I'm afraid he didn't (at least initially) understand what I'm asking for (no offense to him, of course).

To me, this pattern is a great way to achieve maximum performance (both CPU and RAM wise) using most of what language already offers (or is slated to offer soon). The only thing in the way is a restriction which doesn't make much sense and is the opposite of the language rules elsewhere regarding the same subject. Also, the workarounds are either ugly or introduce a lot of boilerplate.

But at least (I hope) now people understood what I'm talking about and maybe some day somebody will champion this.

**Reply by jaredpar** [reply]
> Yes but I'm afraid he didn't (at least initially) understand what I'm asking for (no offense to him, of course).

I zerod in on the `readonly` and `??=` call and got trapped into thinking this was a case for `init` members. The fact that the field is a `ref` field escaped me at first. 

> I mean, you got Jared thinking about this. I personally do not think this is a particularly compelling use case.

An issue with `ref struct`, `readonly` is really not that hard to suck me into 😄

**Reply by jaredpar** [reply]
I think the right starting point for this conversation is identifying concretely why we decided to error on this case. Also potentially what other cases we might be erroring on that are related to this. I took a look at the spec to see what it says here and I'm actually having trouble finding the line which says this case should be illegal. The best I could find is 

> A member initializer that specifies a collection initializer after the equals sign is an initialization of an embedded collection. Instead of assigning a new collection to the field or property, the elements given in the initializer are added to the collection referenced by the field or property. The field or property must be of a collection type that satisfies the requirements specified in §7.6.10.3

The referenced section says that ...

> For each specified element in order, the collection initializer invokes an Add method on the target object with the expression list of the element initializer as argument list, applying normal overload resolution for each invocation. Thus, the collection object must contain an applicable Add method for each element initializer.

Let's take the original sample here and reduce it a bit to the core problem: 

```csharp

using System;
using System.Collections;

// User writes
var foo = new Foo
{
  // CS1918
  Numbers = { 123 }
};

// Spec says this is what is generated
var __temp = new Foo();
// No error
__temp.Numbers.Add(123);
var foo = __temp;

class Foo
{
  public Wrapper Numbers => throw null!;
}
public struct Wrapper : IEnumerable 
{
  public void Add(int i) { }
  IEnumerator IEnumerable.GetEnumerator() => throw new NotSupportedException();
}
```

The _generated_ code compiles without error which, according to the spec, should mean that it's legal. Not sure why `readonly` needs to come into play at all here.

Given that before taking any new changes I would want to understand concretely why do we error here at all. Nothing I found in the spec says it should error and the fact that the expanded out version doesn't error deepens my suspicions.

**Reply by TahirAhmadov** [reply]
This divergence in outcomes is exactly what I was talking about a few comments up. This CS1918 error is specifically raised inside of object initializers as a special rule for that context. The fact that the actual transpiled code is legal illustrates how disconnected these 2 sets of rules are. I'm willing to bet that when object initializers were added, the team responsible for them decided in their own "silo" that the risk of subtle bugs warrants outright prohibition; without consulting the people who decided that doing so outside of object initializers is fine and we just emit a silent copy and call it a day.

This, or there is some very complicated reason all of us are missing at this point.

The reason I brought up `readonly` is that since it was added to the language, these rules can be made more intelligent. I totally understand why simple property setting would be illegal - because the effect of those setters would be immediately lost. However, `readonly` members are a great way for the author to indicate to the compiler that this is a different situation and the author is doing something else other than simply modifying the struct, and therefore invoking such member (be it setting a property or adding a collection item) should be legal.

**Reply by TahirAhmadov** [reply]
In fact, the collection isn't even the important thing here - just regular property setting is prohibited the same way.

[SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAICYCMBYAKEIDcIAnAAgGMKBeCmAUwHcKBhQgb0IooGU6FbgV4UACmQD2AB0EAiCMCpyeFAL6E1AbkKEMAZgrZ2Q1bwP9LnCgHNGCLRQDO9xxoLuLThGQCuVBH4uMyNDDBwABnEpWU47B2dXdU0gA=)

```cs
var c = new C
{
  S = { // CS1918
   Prop = "abc"
  }
};

public class C {
    public S S { get; set; }
}
public struct S
{
    public string Prop {get; set; }
}
```

**Reply by timcassell** [reply]
> In fact, the collection isn't even the important thing here - just regular property setting is prohibited the same way.

Well, that case makes sense, because it's modifying a copy. However, I tried changing it to return `ref S` and it still errors.

**Reply by jaredpar** [reply]
> This, or there is some very complicated reason all of us are missing at this point.

There are a number of places in the language where we disallow likely mutating operations on struct r-values. The rationale being that this is most likely a bug, and enough of one, that it should be disallowed. For instance assigning into properties of a struct r-value. Without actually digging into the code I would _suspect_ that the reason for this rule.

> The fact that the actual transpiled code is legal illustrates how disconnected these 2 sets of rules are. I'm willing to bet that when object initializers were added, the team responsible for them decided in their own "silo" that the risk of subtle bugs warrants outright prohibition; without consulting the people who decided that doing so outside of object initializers is fine and we just emit a silent copy and call it a day.

The compiler team, even back in the time of collection initializers, rarely made decisions like this without consulting with language design. I think it's more likely that during implementation this error was added for some reason and it just wasn't recorded into the spec. Can find plenty of cases where this happened over the years. 

I think it's important to note that the tight relationship between proposals and implementations is a relatively new event in C#. The last few years it's quite common to see PRs in roslyn be paired up with PRs into csharplang to keep the spec and compiler in lock step. That is great because it means our specs are much more closely aligned with the implementation. But go back even say ~3-4 years and you won't see a strong of a pairing even though it was all on GitHub. 

I do think the next steps for this proposal is a bit of code archeology. Basically dig through the test cases that have this error code and see if they, or the code which raises it, have any comments or hint as to why it was added. If it is just a struct r-value problem then the path forward may be as simple as deciding to remove that restriction.

**Reply by TahirAhmadov** [reply]
@timcassell can you post your ref code so that we can see exactly what you mean?

**Reply by TahirAhmadov** [reply]
@jaredpar i agree that if we can find evidence - why it was done, why it's different than elsewhere, etc - we should definitely take that into account when deciding what to do here. I'm willing to look into it if you can provide a link to the test cases repo/ folder.

I wouldn't, however, lift the restriction outright. I do see the value in at least warning about invoking non readonly members of structs like this.

**Reply by timcassell** [reply]
> @timcassell can you post your ref code so that we can see exactly what you mean?

```cs
var c = new C
{
  // CS1918
  S = {
   Prop = "abc"
  }
};
public class C {
    private S _s;
    public ref S S => ref _s;
}
public struct S
{
    public string Prop {get; set; }
}
```

**Reply by TahirAhmadov** [reply]
@timcassell that looks like yet another "flaw" - it should be legal. Most likely, this was never brought up by anybody or discussed in any way.

**Reply by jaredpar** [reply]
> I do see the value in at least warning about invoking non readonly members of structs like this.

The generated code does not warn though. Why should the init syntax warn?

**Reply by TahirAhmadov** [reply]
I think the generated code should warn, too. 
I understand that would be a breaking change, but for init blocks we're relaxing, so it's not a breaking change.
Having said that, I'm OK with full "legalization" - right now the error is completely prohibitive.

**Reply by jaredpar** [reply]
To warn for the generated code you'd roughly need to warn whenever a method was called a on a `struct` that was an r-value. That would be an incredible amount of call sites. I doubt that LDM would consider that level of a breaking change. 

The reason why I'm pushing on this is I don't really see how `readonly` is a factor here. Maybe if we were designing the language from scratch where we had `readonly` from the start we could take the position that calling non-readonly methods from a r-value is bad and we should warn. That would force `readonly` to be embraced by a lot more customers but if you're doing that from the start it's doable. At this point though that moment has passed. Have to accept that calling non-readonly methods on rvalues is just part of the language.

**Reply by TahirAhmadov** [reply]
Regarding `readonly` and "from scratch" question, maybe it's just me but I always like to start by imagining what would be the best possible solution (given current knowledge, of course) and thinking about a way to achieve that solution. Call me a "breaking change hawk", if you will. It's better to spend to pay back the debt as soon as possible, in my book. I do understand that a language is much more tricky than "regular" software, but you guys recently explicitly stated that you are going to consider potential breaking changes when they're justified.
Having said that, I understand what you mean and I wrote everything there is to write on this (especially in #8355). There is no point in continuing to debate that topic. At the end of the day, an analyzer is straightforward to write for this.

The other question - should init blocks follow the same exact rule - hinges on whether there is an overarching "law" that the init block rules have to match the rules of the generated code. I don't know whether such a "philosophy" exists or not and will take your word for it. 
- If such a "law" exists, then I'm perfectly fine with full "legalization" in init blocks, and then I can make/use an analyzer to handle both cases, to make my code better.
- If it doesn't, I would argue that this is a good place to have a divergence, especially with the (however slim) possibility that at some point in the future, the rule outside of init blocks can be tightened (imagine, an analyzer becomes popular and it's an industry-wide agreement that this rule has to be part of the language, and you have multiple customers requesting this.)

**Reply by jaredpar** [reply]
> ... , maybe it's just me but I always like to start by imagining what would be the best possible solution (given current knowledge, of course) and thinking about a way to achieve that solution

How is the best possible solution making the same method invocation behave differently in the same scenario? Or pushing 20 years of .NET ecosystem around mutable `struct` to be forced to understand `readonly`? Consider that there are likely loads of `struct` based collections out there today that could work with collection initializers but don't because of this restriction. But if you constrain the solution to only `readonly` methods then those types are still blocked. How is this the best possible solution? 

> Call me a "breaking change hawk", if you will. It's better to spend to pay back the debt as soon as possible, in my book

There is no debt here. The only debt is the desire of a _very small minority_ of users on this repo and the discord channel. The overwhelming majority of C# customers (99%+) do not care about `readonly`, the breaking changes that people want to introduce around them. They have 20 years of writing `struct` without `readonly` and are quite happy to do so. Conversations around broad sweeping changes like warnings on silent struct copy or readonly correctness is not moving the language design forward. It is just taking up time that could be better spent discussing practical solutions the language could actually implement.

**Reply by TahirAhmadov** [reply]
@jaredpar  

> But if you constrain the solution to only `readonly` methods then those types are still blocked.

I'm only advocating for warnings, and those can be suppressed. So it's not correct to characterize that as "blocked".

> There is no debt here.

One last time - allowing invoking modifying members on r-value structs which are immediately lost is a "pit of failure". There is literally no reason to do it and every reason not to do it. And, at least somebody on the C# team at some point had the same opinion, so much so that they made it an error - CS1918 - not just a warning.
PS. I understand the argument that a lot of these members are "readonly" in nature despite not being decorated as such, but outside of init blocks, we're talking about mostly methods and I can imagine people have been writing "no effect" methods on structs, whereas in init blocks, we would be completely unblocking setting properties and fields and those are much more likely to be modifying (and the field setting is definitely modifying).

Regarding practical solutions, I'd be happy with full removal of CS1918, while having a preference for partial removal with warnings for non-readonly members. 
As a good compromise, we can completely remove CS1918 and make an analyzer available under the Microsoft namespace for those who want to improve their code hygiene - I'm willing to write said analyzer.

**Reply by jaredpar** [reply]
> I'm only advocating for warnings, and those can be suppressed. So it's not correct to characterize that as "blocked".

Warnings are breaking changes. Particularly when the scope of the warning is significant portions of the .NET ecosystem. This is a non-starter. 

Producing a warning where we feel like the majority response will be to suppress de-values warnings. It gets customers into a pattern of feeling okay suppressing warnings on upgrade. That is not a behavior we want to encourage. Today we issue warnings because issuing an error isn't really possible after certain period of time. We want to get customers attention, we want them to consider their code and take action. 

> One last time - allowing invoking modifying members on r-value structs which are immediately lost is a "pit of failure"

There is no way to differentiate between when a member is or isn't modifying the struct value directly. The use of `readonly` is an assertion that it's not (but it's not a guarantee). The absence of `readonly` is not a guarantee or even affirmative statement that the modification is lost. It is just a statement that this is how structs have been authored for 20 years. Attaching a warning here is satisfying a _very small minority_ of customers at the expense of a large set of customers revisiting code that works perfectly well today. 

It is very hard to maintain interest in these proposals when we are revisiting this topic over and over again.

**Reply by TahirAhmadov** [reply]
> Warnings are breaking changes.

Outside of init blocks - yes, it's a breaking change, but inside init blocks, it's currently an error, so it's a relaxation of rules.
Like I said, I'm not going to argue for the breaking change outside of init blocks anymore; there is a separate discussion for that and there is no point to keep going in circles.

> We want to get customers attention, we want them to consider their code and take action.

Same idea here - mark those members `readonly` and the warning goes away.

> The use of readonly is an assertion that it's not (but it's not a guarantee).

Can you elaborate why it's not a guarantee?

>  The absence of readonly is not a guarantee or even affirmative statement that the modification is lost.

- For fields, those set values are guaranteed to be lost. 
- For properties, they are not guaranteed, but are much more likely than not to be lost in most code (unless there is evidence there are a large number of structs with full body properties which make changes outside of the struct?) 
- For methods, those are outside of init blocks, and this discussion is about relaxing/removing CS1918, not tightening the general rule.

PS. The first two bullet points above about fields and props are arguments for why inside init block, (not in general!) we should go from error CS1918 to warning for non-readonly members (fields and non-readonly properties, and non-readonly Add method for collections). But again, I'll be fine with full removal of CS1918, as well.

**Reply by TahirAhmadov** [reply]
To summarize:
- The general rule discussion is #8355; I'm not going to debate that here anymore.
- The init block discussion is here, the removal of CS1918. I'm happy with either
  - full removal of CS1918, (potentially with a MS-shipped analyzer for those who want it) or 
  - with partial relaxation with warnings etc.

**Reply by jaredpar** [reply]
> Same idea here - mark those members readonly and the warning goes away.

Customers do not own all the code they consume. Telling them just mark it `readonly` means a substantial amount of customers cannot follow your advice. 

> Can you elaborate why it's not a guarantee?

There are a number of ways that a `readonly` member can mutate `this`. 

> For fields, those set values are guaranteed to be lost.

Incorrect. When the fields are reference types, struct backed by reference type, using mutable statics, etc ... then the mutations persist. 

> For properties, they are not guaranteed, but are much more likely than not to be lost in most code 

Properties are just method calls by another name. They are no different in terms of capabilities here. 

> For methods, those are outside of init blocks, and this discussion is about relaxing/removing CS1918, not tightening the general rule.

You insist on bringing `readonly` into this discussion. That means you have to discuss the behavior of method calls outside of collection initilaizers because that is exactly how the behavior is specified.

**Reply by TahirAhmadov** [reply]
> Customers do not own all the code they consume. Telling them just mark it readonly means a substantial amount of customers cannot follow your advice.

For init blocks, previously it was an error altogether, so I'm not sure why it's such a big deal that they have to now deal with a warning.

> There are a number of ways that a readonly member can mutate this.

Sure, but those are "hackish". I'm not saying it can be made airtight, nor should we try.

> Incorrect. When the fields are reference types, struct backed by reference type, using mutable statics, etc ... then the mutations persist.

I meant setting those fields, not mutating referenced objects. The latter should be allowed, ditto for `ref`s for structs.

> Properties are just method calls by another name. They are no different in terms of capabilities here.

I know that 🤣 However, just common sense tells us that 99% of property setters modify the struct itself, either it's an auto property, or it sets a field, etc. That's why i said - if there are many properties with outside-of-struct side effects, I would like to see evidence of that - I just don't think that's the case. And again, currently all this is an error anyway, so....

> You insist on bringing `readonly` into this discussion. That means you have to discuss the behavior of method calls outside of collection initilaizers because that is exactly how the behavior is specified.

Nope - I mention `readonly` for property setters and `Add` methods in the context of init block collection item adders, not regular method calls outside of init blocks. Again, I'm not debating the general rule outside of init blocks. I'm **not**. 🤣

**Reply by TahirAhmadov** [reply]
Once more - if you think full removal of CS1918 and completely bringing the init blocks behavior inline with the general rule outside of init blocks is the way to go, great. I'll take that and be a happy guy. 👍

**Comment by timcassell** [comment]
It's a different error, but it seems very closely related to #9174. @jnm2

[Edit] Ah, I see it was called out in #9175.

**Reply by jnm2** [reply]
Nice to find that it's not purely theoretical! Thanks for the crosslink making me aware of this thread.

**Comment by TahirAhmadov** [comment]
Any chance somebody wants to champion this? It's really becoming a blocker. The most compelling use case is: imagine a UI element tree, and in these cases the elements have child elements, as well as other collections (depending on element type). All of these are **usually** initialized once and stay the same for the running time of the application. (Sometimes UI elements can be added/removed, but those situations are rare.) The most efficient storage mechanism would be something like:
```cs
struct ListStructCompact<T> : IEnumerable<T>
{
  public ListStructCompact() { this._items=[]; }
  T[] _items;
  void Add(T item) { ... } // resize _items by one size up
  // other boilerplate
}
class Control
{
  public ref ListStructCompact<Control> Children => ref this._children;
  ListStructCompact<Control> _children = [];
}

this.Controls.Add(new Panel
{
  Children = { new Button { } }, // error - CS1918
});
```

Unlike other syntax sugar language improvements, this one is an actual blocker. To use the object init block collection add feature, I have to make it a class type, which necessitates something like List<T>, which is 2 allocations - the List<T> and the array (as opposed to only one, the array, as it is with ListStructCompact<T>). 
Further, List<T> does exponential growth of the array - yes, it's easy enough to fix with a custom ListCompact<T> which grows the array by one element, but it's a half-solution - I don't want to spend the time to refactor the code just for this improvement, without also doing the struct one.
Ideally, I can use ListStructCompact<T> and also we get something like `Children += [new Button { }, new TextBox { }]`. This would make this situation perfect: no unneeded allocations; no unused array space; no need for an extra `int _size` field; and in 99% of the cases - the initial construction of element/control tree - arrays are not resized multiple times, because the ` += [...]` can call AddRange behind the scenes.
I heard that there is talk about making C# better for UI stuff - this should be the top improvement, IMO.

**Reply by jnm2** [reply]
I'm championing and this feature is already approved for Any Time implementation (links available above and below your comment)

**Reply by TahirAhmadov** [reply]
Thank you!

**Comment by jnm2** [comment, marked as answer]
This feature is approved for Any Time implementation, as an approved expansion within https://github.com/dotnet/csharplang/issues/9174!
